File: listas/listas.py
import random

# Crea un funcion para imprimir una lista
def printList(values):

    print("[ ", end = "")

    # recorre la lista
    for x in values:

        print(x, end = " ")

    print("]")

# funcion para aplicar un bubblesort
def bubbleSort(values):
    
    lenVar = len(values)

    # recorre la lista 
    for i in range(lenVar):

        # recorre la lista menos el ultimo elemento
        for j in range(lenVar - 1):

            # si el valor en j es mayor al de j + 1 los intercambia
            if values[j] > values[j + 1]:
                values[j], values[j + 1] = values[j + 1], values[j]

# funcion para aplicar un selectionsort
def selectionSort(values):

    lenVar = len(values)

    # recorre la lista 
    for i in range(lenVar):

        # toma index como pivote
        minI = i

        # recorre el array para buscar un nuevo valor más bajo desde del ultimo más bajo, i
        for j in range(i, lenVar):

            # si valor actual en j es menor a minI actual y actualiza minI
            if values[j] < values[minI]:

                minI = j

        # los intercambia
        values[minI], values[i] = values[i], values[minI]

#  funcion para aplicar un insertionSort
def insertionSort ( values ):

    lenVar = len(values)

    # recorre la lista desde 1 asumiendo que el primer elemento ya esta ordenado
    for i in range(1, lenVar):

        # si el valor en i es menor al valor anterior en i
        if values[i] < values[i-1]:

            # usa i como pivote
            piv = values[i]

            # crea una variable temporal para i
            temp = i

            # desplaza los valores a la izquierda mientras que no sean 
            # temp 0 para eviar un outOfBounds
            # values[temp - 1] > piv para que no reemplace un valor menor al pivote actual
            while temp > 0 and values[temp-1] > piv:

                # toma el valor anterior y lo reemplaza con el actual
                values[temp] = values[temp - 1]

                temp-=1

            # reemplaza la ubicacion actual de temp con el pivote
            # "lo inserta"
            values[temp] = piv

# Crea una lista
listaPersonalizada = []

# Asgina valores aleatorios 
for i in range(10):

    listaPersonalizada.append(random.randint(1, 8))

print("Lista desordenada:", listaPersonalizada)
# Se aplica el bubbleSort
bubbleSort(listaPersonalizada)
print("Lista ordenada con bubbleSort:", listaPersonalizada)
print("=====")

# Asigna valores aleatorios
for i in range(10):
    listaPersonalizada[i] = random.randint(1, 8)

print("Lista desordenada:", listaPersonalizada)
# Se aplica el selectionSort
selectionSort(listaPersonalizada)
print("Lista ordenada con selectionSort:", listaPersonalizada)
print("=====")

# Asigna valores aleatorios
for i in range(10):
    listaPersonalizada[i] = random.randint(1, 8)

print("Lista desordenada:", listaPersonalizada)
# Se aplica el insertionSort
insertionSort(listaPersonalizada)
print("Lista ordenada con insertionSort:", listaPersonalizada)
print("=====")

# Asigna valores aleatorios
for i in range(10):
    listaPersonalizada[i] = random.randint(1, 8)

print("Lista desordenada:", listaPersonalizada)

# Elimina el primer elemento de la lista
listaPersonalizada.pop(0)

# Obtiene el cuarto elemento de la lista
listaPersonalizada.index(3)

# Organiza los elementos de la lista
listaPersonalizada.sort()

print("Lista ordenada por el metodo del SDK, luego de agregar y eliminar valores:", listaPersonalizada)

File: listas/listas.cpp
#include <stdio.h>
#include <cstdlib>
#include <ctime>
#include <vector>
#include <algorithm>

using namespace std;

// Crea una funcion para imprimir un vector int
void printVector( vector<int> values ) {

    printf("[ ");
    
    // Recorre el array
    for ( int i = 0; i < values.size(); i++ ) {

        printf("%i, ", values[i]);

    }

    printf("]\n");
    
}

// funcion para generar numero aletatorio
int randomInt( int min, int max ) {

    // retorna un numero aleatorio
    return rand() % max + 2 - min ;

}

// funcion para aplicar un bubblesort
void bubbleSort( vector<int>& values ) {

    // recorre el array
    for ( int index = 0; index < values.size(); index++ ) {
        
        // recorre el arrya menos el ultimo elemento
        for ( int index2 = 0; index2 < values.size() - 1; index2++ ) {

            // si el valor en index2 es mayor al de index2 + 1 los intercambia
            if ( values[index2] > values[index2+1] ) {

                int temp = values[index2];

                values[index2] = values[index2+1];

                values[index2+1] = temp;
                
            }
            
        }

    }

}

// funcion para aplicar un selectionsort
void selectionSort( vector<int>& values ) {

    // recorre el array
    for ( int i = 0; i < values.size(); i++ ) {

        // toma index como pivote
        int minI = i; 

        // recorre el array para buscar un nuevo valor más bajo desde del ultimo más bajo, i
        for ( int j = i; j < values.size(); j++ ) {

            // si valor actual en j es menor a minI actual y actualiza minI
            if ( values[j] < values[minI] ) {

                minI = j;

            }

        }

        // los intercambia
        int temp = values[minI];

        values[minI] = values[i];

        values[i] = temp;

    }

}

// funcion para aplicar un insertionsort
void insertionSort( vector<int>& values ) {

    // recorre el array asumiendo que el primer valor ya está ordenado
    for ( int i = 1; i < values.size(); i++ ) {
        
        // si el valor en i es menor al valor anterior en i
        if ( values[i] < values[i - 1] ) {
            
            // usa i como pivote
            int piv = values[i];
            
            // crea una variable temporal para i
            int temp = i;
            
            // desplaza los valores a la izquierda mientras que no sean 
            // temp 0 para eviar un outOfBounds
            // values[temp - 1] > piv para que no reemplace un valor menor al pivote actual
            while ( temp > 0 && values[temp - 1] > piv ) {
                
                // toma el valor anterior y lo reemplaza con el actual
                values[temp] = values[temp - 1]; 
                
                temp--;
                
            }
            
            // reemplaza la ubicacion actual de temp con el pivote
            // "lo inserta"
            values[temp] = piv;
            
        }
    }

}

int main() {

    // semilla de generacion de numeros
    srand(time(NULL));

    // Crear una lista (vector)
    vector<int> listaPersonalizada(10);

    // Asgina valores aleatorios
    for (int i = 0; i < 10; i++) {

        listaPersonalizada[i] = randomInt(1, 8);

    }

    printf("Lista desordenada: ");
    printVector(listaPersonalizada);
    // Se aplica el bubbleSort
    bubbleSort(listaPersonalizada);
    printf("Lista ordenada con bubbleSort: ");
    printVector(listaPersonalizada);
    printf("=====\n");

    // Asgina valores aleatorios 
    for (int i = 0; i < 10; i++) {
        listaPersonalizada[i] = randomInt(1, 8);
    }

    printf("Lista desordenada: ");
    printVector(listaPersonalizada);
    // Se aplica el selectionSort
    selectionSort(listaPersonalizada);
    printf("Lista ordenada con selectionSort: ");
    printVector(listaPersonalizada);
    printf("=====\n");

    // Asgina valores aleatorios 
    for (int i = 0; i < 10; i++) {
        listaPersonalizada[i] = randomInt(1, 8);
    }

    printf("Lista desordenada: ");
    printVector(listaPersonalizada);
    // Se aplica el insertionSort
    insertionSort(listaPersonalizada);
    printf("Lista ordenada con insertionSort: ");
    printVector(listaPersonalizada);
    printf("=====\n");

    // Asgina valores aleatorios 
    for (int i = 0; i < 10; i++) {
        listaPersonalizada[i] = randomInt(1, 8);
    }

    printf("Lista desordenada: ");
    printVector(listaPersonalizada);

    // Elimina el primer elemento de la lista
    listaPersonalizada.erase(listaPersonalizada.begin());

    // Obtiene el cuarto elemento de la lista
    int index = distance(listaPersonalizada.begin(), find(listaPersonalizada.begin(), listaPersonalizada.end(), 3));

    // Organiza los elementos de la lista
    sort(listaPersonalizada.begin(), listaPersonalizada.end());

    printf("Lista ordenada por metodo del SDK, luego de agregar y elimnar valores: ");
    printVector(listaPersonalizada);

    return 0;
}
